### DISCUSSION TITLE
Is there a security risk from exposing PockerBase directly to the internet?

### DISCUSSION THREAD

**Discussion by takinola** [pocketbase/pocketbase#3297] [Q&A] [answered]
Exposing Pocketbase behind a reverse proxy means that there will be a web accessible login form to the database.  Theoretically, an attacker could bruteforce the admin access by trying every combination of user names and passwords given the speed of sqlite.  This is one of the reasons when working other databases (MySQL, Redis, etc), it is advised to bind access to only localhost.

Does PocketBase implement any protections against an attack like this eg throttling abusive users, computationally expensive hashing, etc?

Another option would be to remove the need to expose PocketBase to the internet by enabling all configuration to be done by API or configuration files and removing the need for an admin UI in the first place.

**Comment by ganigeorgiev** [comment, marked as answer]
The passwords are hashed and verified with `bcrypt` and the speed of each operation is with constant time and will depend on the hash cost/weight/iteration factor (_we currently use `12`_) and doesn't depend on the database.

There is no rate limiting or throttling at the moment but as long as you use complex passwords (and don't reuse them!) I wouldn't worry that much about bruteforce attacks.

If you really need it, you can configure rate-limiting on reverse proxy (nginx, apache, caddy, etc.) level.
Alternatively, on reverse proxy level you can also whitelist only your IP(s) to access the `/api/admins/*` endpoints (_note that the Admin UI is just a static client application that uses the exposed web APIs_).
